File: app/src/main/java/com/practice/forecast/ui/map/WeatherViewModelFactory.java
package com.practice.forecast.ui.map;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

public class WeatherViewModelFactory implements ViewModelProvider.Factory {

    private final ViewModel weatherViewModel;

    public WeatherViewModelFactory(ViewModel weatherViewModel) {
        this.weatherViewModel = weatherViewModel;
    }

    @SuppressWarnings("unchecked")
    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        ViewModel viewModel;
        if(modelClass == WeatherMapViewModel.class){
            viewModel = weatherViewModel;
        } else {
            throw new RuntimeException("unsupported view model class: " + modelClass);
        }
        return (T) viewModel;
    }
}

File: app/src/main/java/com/practice/forecast/ui/map/WeatherViewModelFactory.kt
package com.practice.forecast.ui.map

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider

class WeatherViewModelFactory(private val weatherViewModel: ViewModel) : ViewModelProvider.Factory {
    override fun <T : ViewModel?> create(modelClass: Class<T>): T {
        val viewModel: ViewModel
        viewModel = if (modelClass == WeatherMapViewModel::class.java) {
            weatherViewModel
        } else {
            throw RuntimeException("unsupported view model class: $modelClass")
        }
        return viewModel as T
    }
}